### ISSUE TITLE
launching ddp fails with assertion error

### ISSUE THREAD

**Issue by awaelchli** [PyTorchLightning/pytorch-lightning#8998]
## 🐛 Bug

Originally reported here: https://github.com/PyTorchLightning/pytorch-lightning/issues/4612#issuecomment-901539844

same for me, but the code doesn't get stuck at commit 03bb389 so it must be a change introduced afterwards

@williamFalcon (idk who else to tag) I tried to track down the problem by comparing the old commit and master.

```
plugins/training_type/ddp.py:131

     def setup_environment(self) -> None:
         # start the other scripts
-        if not self.cluster_environment.creates_children():
+        if not self.cluster_environment.creates_children() and os.environ.get("PL_IN_DDP_SUBPROCESS", "0") != "1":
             self._call_children_scripts()
```
removing os.environ... makes all created subprocesses go into _call_children_scripts() and then fail with assert local_rank == 0 causing the block, a solution would be reinstating the os check or adding:

```
if not self.cluster_environment.creates_children() and self.local_rank == 0:
    self._call_children_scripts()
```

### To Reproduce

TODO

### Environment

- PyTorch Lightning Version (e.g., 1.3.0):
- PyTorch Version (e.g., 1.8)
- Python version:
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

**Comment by awaelchli** [comment]
@sordonia would you kindly provide the following information:

- command how you launch ddp from the terminal
- output of print(trainer.accelerator_connector.cluster_environment)
- optionally a running minimal [debug model](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py):

**Comment by sordonia** [comment]
Thanks for following up:

1. in 1.5.0, LightningEnvironment assigns:

```
def creates_children(self) -> bool:
    # return True if the cluster is managed (you don't launch processes yourself)
    return 'LOCAL_RANK' in os.environ
```

my backend (AzureML) assigns LOCAL_RANK even if it doesn't create children (LOCAL_RANK=0) when the job starts in a multi-gpu machine with 1 process. Evidently, if I just run it out-of-the box my code get stuck given that creates_children returns True when actually it is not True.

2. If LOCAL_RANK=0 is set, I create an AzureEnvironment as follows:

```
class AzureClusterEnvironment(LightningEnvironment):
    def creates_children(self) -> bool:
        # return True if the cluster is managed (you don't launch processes yourself)
        return False
```

and pass it with `plugins=[AzureClusterEnvironment()]` to Trainer. In this case, my job fail with `assert self.local_rank == 0`. It seems that the other processes indeed go into `self.create_children()`. To reproduce:

1. export LOCAL_RANK=0
2. modify Trainer by just passing `plugins=[AzureClusterEnvironment()]`
3. python boring_model.py

**Comment by awaelchli** [comment]
> my backend (AzureML) assigns LOCAL_RANK even if it doesn't create children (LOCAL_RANK=0) when the job starts in a multi-gpu machine with 1 process. 

that's not very good of AzureML. this environment variable should not just "be there" and obviously it will affect all processes running. To avoid your issues with multi-gpu, you should unset it. In bash shell:

```
unset LOCAL_RANK
```

However, you are mentioning a single GPU use case too, and there it should not matter. In Lightning, even if you set accelerator=ddp it will not spawn processes unless gpus > 1 (obviously).  The assertion error you are mentioning cannot happen unless local rank would actually be > 1, but this is actually happening because you set creates_children=False constant, so even the child processes will go and try to spawn again. 

You could create an azure ml environment like this one:

```python
import os

class AzureClusterEnvironment(LightningEnvironment):
    def __init__(self):
        super().__init__()
        del os.environ["LOCAL_RANK"]
```

Does my explanation make sense?

**Comment by sordonia** [comment]
It makes sense thanks !

**Comment by awaelchli** [comment]
@sordonia [Here](https://github.com/PyTorchLightning/pytorch-lightning/issues/8998#issuecomment-901982631) I asked for how you launch the code and you answered you launch like this: python boring_model.py
There is no openmpi involved here. If Lightning gets invoked like this, it will launch as many processes as you request (gpus>0). 

However, then you wish to launch with openmpi. That's also fine. If you do that and openmpi assigns LOCAL_RANK to the process, then Lightning will recognize that and NOT launch new processes as this is the responsibility of the mpi backend. 
So I see no conflicts here. 

In summary: 
- LOCAL_RANK present -> Lightning assumes user launches processes. A launch utility is needed, or user typing the commands manually in the terminal. 
- LOCAL_RANK not present -> Lightning assumes user wants N processes, so Lightning will launch N-1 additional ones which will have LOCAL_RANK=1, ... N-1 set.

**Comment by sordonia** [comment]
Thanks, yes indeed, there is no problem if I launch with python boring_model.py locally as OpenMPI is not used.

In AzureML, where OpenMPI is used, you have two options for distributed jobs. Say that you have 2 nodes at your disposal with 4 gpus each. OpenMPI will be used to either:

1. launch 2 processes, one per node, and set LOCAL_RANK=0 on each process (it's the only process)
2. launch 8 processes, 4 per node, and set LOCAL_RANK=0,1,2,3 on each process.

PL seems to rely on the assumption that in case 1. LOCAL_RANK is not set , instead of being set to 0. Is this reasonable? Case 1 seems just a particular case of the general rule.

Thanks

**Comment by sordonia** [comment]
cool thanks!
